core/src/main/java/org/apache/calcite/sql/SqlDialect.java

```diff
@@ -1514,9 +1514,15 @@ public static Pair<List<String>, List<List<Character>>> getDateTimeTokensWithSep
   private static boolean isDotSeparatorInAMPM(
       Character currentChar, String standardDateFormat, int indexofCurrentChar) {
     return currentChar.toString().equals(".")
-        && (standardDateFormat.charAt(indexofCurrentChar - 1) == 'A'
-            || standardDateFormat.charAt(indexofCurrentChar - 1) == 'P'
-            || standardDateFormat.charAt(indexofCurrentChar - 1) == 'M');
+        && (
+            (standardDateFormat.charAt(indexofCurrentChar - 1) == 'A'
+        && standardDateFormat.charAt(indexofCurrentChar + 1) == 'M')
+            || (standardDateFormat.charAt(indexofCurrentChar - 1) == 'P'
+        && standardDateFormat.charAt(indexofCurrentChar + 1) == 'M')
+            || (standardDateFormat.charAt(indexofCurrentChar - 1) == 'M'
+        && standardDateFormat.charAt(indexofCurrentChar - 2) == '.'
+        && (standardDateFormat.charAt(indexofCurrentChar - 3) == 'A'
+                || standardDateFormat.charAt(indexofCurrentChar - 3) == 'P')));
```

Comment:
This method was incomplete and was affecting formats which do not have A.M. or P.M.
The current change makes it complete and will return true only in case of formats containing  A.M. or P.M.